File: src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

public class Animal implements WorldElement{
    private MapDirection orientation;
    private Vector2d position;

    public Animal(Vector2d initialPosition) {
        this.orientation = MapDirection.NORTH;
        this.position = initialPosition;
    }

    public Animal() {
        this (new Vector2d(2, 2));
    }

    public MapDirection getOrientation() {
        return this.orientation;
    }

    public Vector2d getPosition() {
        return this.position;
    }

    public String toString() {
        return this.orientation.toString();
    }

    public boolean isAt(Vector2d position) {
        return this.position.equals(position);
    }

    public void move(MoveValidator validator, MoveDirection direction) {
        switch (direction) {
            case FORWARD, BACKWARD -> {
                Vector2d unitVector = this.orientation.toUnitVector();
                if (direction == MoveDirection.BACKWARD) {
                    unitVector = unitVector.opposite();
                }
                Vector2d newPosition = this.position.add(unitVector);
                if (validator.canMoveTo(newPosition)) {
                    this.position = newPosition;
                }
            }
            case RIGHT -> this.orientation = this.orientation.next();
            case LEFT -> this.orientation = this.orientation.previous();
        }
    }
}

File: kotlin/src/main/kotlin/model/Animal.kt
package model

class Animal(private var orientation: MapDirection = MapDirection.NORTH, private var position: Vector2d = Vector2d(2,2)) {

    override fun toString() = orientation.toString()

    fun isAt(position: Vector2d) = this.position == position

    fun getPosition() = this.position

    fun setPosition(position: Vector2d) {
        this.position = position
    }

    fun move(direction: MoveDirection) = when(direction) {
        MoveDirection.RIGHT -> this.orientation = orientation.next()
        MoveDirection.LEFT -> this.orientation = orientation.previous()
        MoveDirection.FORWARD -> this.position += this.orientation.toUnitVector()
        MoveDirection.BACKWARD -> this.position -= this.orientation.toUnitVector()
    }
}